docs/spec/lightclient/detection/req-ibc-detection.md

```diff
@@ -0,0 +1,296 @@
+	  
+2. the relayer may also detect a fork by observing that the
+   handler for chain A (on chain B)
+   is on a different branch than the relayer
+
+
+- in both detection scenarios, the relayer should submit evidence to 
+  full nodes of chain A where there is a fork. As we assume a fullnode
+  has a complete list of blocks, it is sufficient to send "Bucky's
+  evidence" (https://github.com/tendermint/tendermint/issues/5083),
+  that is, 
+     - two lightblocks from different branches + 
+	 - a lightblock (perhaps just a height) from which both blocks 
+	   can be verified.
+  
+- in the scenario 2., the relayer must feed the A-handler (on chain B)
+  a proof of a fork on A so that chain B can react accordingly
+  
+### Handler requirement
+  
+- there are potentially many relayers, some correct some faulty
+
+- a handler cannot trust the information provided by the relayer, 
+  but must verify 
+  (Доверя́й, но проверя́й)
+
+- in case of a fork, we accept that the handler temporarily stores
+  headers (tagged as verified). 
+  
+- eventually, a handler should be informed
+ (`checkMisbehaviourAndUpdateState`)
+ by some relayer that it has
+  verified a header from a fork. Then the handler should do what is
+ required by IBC in this case (stop?)
+
+### Challenges in the handler requirement
+
+- handlers and relayers work on different lightstores. In principle
+  the lightstore need not intersect in any heights a priori
+
+- if a relayer  sees a header *h* it doesn't know at a handler (`queryChainConsensusState`), the
+  relayer needs to
+  verify that header. If it cannot do it locally based on downloaded
+  and verified (trusted?) light blocks, it might need to use
+  `VerifyToTarget` (bisection). To call `VerifyToTarget` we might keep
+  *h* in the lightstore. If verification fails, we need to download the
+  "alternative" header of height *h.Height* to generate evidence for
+  the handler.
+  
+- we have to specify what precisely `queryChainConsensusState`
+  returns. It cannot be the complete lightstore. Is the last header enough?
+
+- we would like to assume that every now and then (smaller than the
+  trusting period) a correct relayer checks whether the handler is on a
+  different branch than the relayer. 
+  And we would like that this is enough to achieve
+  the Handler requirement.
+  
+   - here the correctness argument would be easy if a correct relayer is
+     based on a light client with a *trusted* state, that is, a light
+     client who never changes its opinion about trusted. Then if such a
+     correct relayer checks-in with a handler, it will detect a fork, and
+     act in time.
+
+   - if the light client does not provide this interface, in the case of
+     a fork, we need some assumption about a correct relayer being on a
+     different branch than the handler, and we need such a relayer to 
+	 check-in not too late. Also
+     what happens if the relayer's light client is forced to roll-back
```

Comment:
I guess this scenarios assumes fork detected by the light client. In that case it's game over and the only left to be done is evidence creation and submission. So we need to spec what is happening in the absence of forks (normal operation), then fork detection and evidence submission once fork is detected. But there are no continuation of the normal operation after fork is detected as we don't support automatic recovery.

Reply:
You are right. I had the impression that we might think about rolling back but now I realize that that most likely doesn't make sense.

So if a relayer detects a fork (however it does so), does it have to halt all handlers?

Reply:
It should submit proof of fork to affected handler (if fork is on chain A, it should send it to IBC handler for A on chain B) and this should probably halt handler.